src/lib/form/form-submit-handler.js

```diff
@@ -0,0 +1,120 @@
+import { dispatchCustomEvent } from '../events.js';
+import { isStringEmpty } from '../utils.js';
+import { focusInputAfterError, isCcInputElement, isNativeInputElement, getFormData } from './form-utils.js';
+import { invalid, VALID } from './validation.js';
+
+/**
+ * @typedef {import('./validation.types.js').Validation} Validation
+ * @typedef {import('./form.types.js').HTMLFormElementEvent} HTMLFormElementEvent
+ * @typedef {import('./form.types.js').FormValidation} FormValidation
+ * @typedef {import('./form.types.js').AggregatedFormData} FormSubmittedData
+ */
+
+/**
+ * Creates a function that is to be set as a `<form>` element 'submit' event listener.
+ *
+ * This function performs a programmatic validation according to the constraints set on input elements.
+ *
+ * The validation is made according to the constraints set on input elements.
+ *
+ * When form is invalid, the first invalid input is automatically focused and a `form:invalid` event is dispatched with the validation result.
+ * Error reporting is made on cc input elements: error message will be displayed.
+ * But no error reporting will be made on native input elements because we think that the native behavior has bad UX and has bad accessibility design.
+ *
+ * When form is valid, a `form:submit` event is dispatched with the form data.
+ *
+ * @return {((e: HTMLFormElementEvent) => void)}
+ * @fires {CustomEvent<FormValidation>} form:invalid - Whenever the form is submitted but some inputs are invalid
+ * @fires {CustomEvent<AggregatedFormData>} form:submit - Whenever the form is submitted and all inputs are valid
+ */
+export function formSubmitHandler () {
+  /**
+   * @param {HTMLFormElementEvent} event
+   */
+  return (event) => {
+
+    // we don't want the native form submit
+    event.preventDefault();
+
+    const formElement = event.target;
+
+    // we find the elements that should be validated
+    /** @type {FormValidation} */
+    const formValidation = getFormElementsToValidate(formElement)
+      .map((e) => ({
+        name: e.name,
+        validation: e.validate(),
+      }));
+
+    // now we perform validation on all elements
+    const isFormValid = formValidation.every((result) => result.validation.valid);
+
+    if (isFormValid) {
+      const data = getFormData(formElement);
+
+      dispatchCustomEvent(formElement, 'submit', data);
```

Comment:
**suggestion:** we could rename this to avoid confusion with submit
```suggestion
      dispatchCustomEvent(formElement, 'validation-complete', data);
```
or
```suggestion
      dispatchCustomEvent(formElement, 'error-reporting-complete', data);
```
or
```suggestion
      dispatchCustomEvent(formElement, 'reporting-complete', data);
```